iroha/src/tx.rs

```diff
@@ -1,59 +1,152 @@
-use crate::{crypto::Signature, isi::Contract};
+use crate::prelude::*;
 use iroha_derive::Io;
 use parity_scale_codec::{Decode, Encode};
 use std::time::SystemTime;
 
 /// An ordered set of instructions, which is applied to the ledger atomically.
 #[derive(Clone, Debug, Io, Encode, Decode)]
-pub struct Transaction {
-    /// An ordered set of instructions.
-    //TODO: think about constructor with `Into<Contract>` parameter signature.
-    pub instructions: Vec<Contract>,
-    /// Time of creation (unix time, in milliseconds).
-    creation_time: u128,
-    /// Account ID of transaction creator (username@domain).
-    pub account_id: String,
-    /// Quorum field (indicates required number of signatures).
-    quorum: u32, //TODO: this will almost certainly change; accounts need conditional multisig based on some rules, not associated with a transaction
-    pub signatures: Vec<Signature>,
+pub enum Transaction {
+    Requested {
+        /// An ordered set of instructions.
+        instructions: Vec<Contract>,
+        /// Time of creation (unix time, in milliseconds).
+        creation_time: u128,
+        /// Account ID of transaction creator (username@domain).
+        account_id: Id,
+        signatures: Vec<Signature>,
+    },
+    Accepted {
+        /// An ordered set of instructions.
+        instructions: Vec<Contract>,
+        /// Time of creation (unix time, in milliseconds).
+        creation_time: u128,
+        /// Account ID of transaction creator (username@domain).
+        account_id: Id,
+        signatures: Vec<Signature>,
+    },
+    Signed {
+        /// An ordered set of instructions.
+        instructions: Vec<Contract>,
+        /// Time of creation (unix time, in milliseconds).
+        creation_time: u128,
+        /// Account ID of transaction creator (username@domain).
+        account_id: Id,
+        signatures: Vec<Signature>,
+    },
+    Valid {
+        /// An ordered set of instructions.
+        instructions: Vec<Contract>,
+        /// Time of creation (unix time, in milliseconds).
+        creation_time: u128,
+        /// Account ID of transaction creator (username@domain).
+        account_id: Id,
+        signatures: Vec<Signature>,
```

Comment:
Duplication of internal structure here, maybe use a separate struct?

Reply:
I specially didn't do that because looking for a differences in internal states of variants.
If in the end of implementing validation logic activity they will stay the same - I will extract them into struct.